extensions/guacamole-auth-nextcloud/src/main/java/org/apache/guacamole/auth/nextcloud/user/UserDataService.java

```diff
@@ -0,0 +1,370 @@
+import org.apache.guacamole.auth.nextcloud.ConfigurationService;
+import org.apache.guacamole.auth.nextcloud.RequestValidationService;
+import org.apache.guacamole.net.auth.Connection;
+import org.apache.guacamole.net.auth.Credentials;
+import org.apache.guacamole.net.auth.Directory;
+import org.apache.guacamole.net.auth.User;
+import org.apache.guacamole.net.auth.permission.ObjectPermissionSet;
+import org.apache.guacamole.net.auth.simple.SimpleDirectory;
+import org.apache.guacamole.net.auth.simple.SimpleObjectPermissionSet;
+import org.apache.guacamole.net.auth.simple.SimpleUser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Service for deriving Guacamole extension API data from UserData objects.
+ */
+@Singleton
+public class UserDataService {
+
+    /**
+     * Logger for this class.
+     */
+    private static final Logger logger = LoggerFactory.getLogger(UserDataService.class);
+
+    /**
+     * ObjectMapper for deserializing UserData objects.
+     */
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    /**
+     * Denylist of single-use user data objects which have already been used.
+     */
+    private final UserDataDenylist denylist = new UserDataDenylist();
+
+    /**
+     * Service for retrieving configuration information regarding the
+     * NextcloudJwtAuthenticationProvider.
+     */
+    @Inject
+    private ConfigurationService confService;
+
+    /**
+     * Service for testing the validity of HTTP requests.
+     */
+    @Inject
+    private RequestValidationService requestService;
+
+    /**
+     * Service for handling cryptography-related operations.
+     */
+//    @Inject
+//    private CryptoService cryptoService;
+
+    /**
+     * Provider for UserDataConnection instances.
+     */
+    @Inject
+    private Provider<UserDataConnection> userDataConnectionProvider;
+
+    /**
+     * The name of the HTTP parameter from which base64-encoded, encrypted JSON
+     * data should be read. The value of this parameter, when decoded and
+     * decrypted, must be valid JSON prepended with the 32-byte raw binary
+     * signature generated through signing the JSON with the secret key using
+     * HMAC/SHA-256.
+     */
+    public static final String ENCRYPTED_DATA_PARAMETER = "data";
+
+    /**
+     * Derives a new UserData object from the data contained within the given
+     * Credentials. If no such data is present, or the data present is invalid,
+     * null is returned.
+     *
+     * @param credentials
+     *     The Credentials from which the new UserData object should be
+     *     derived.
+     *
+     * @return
+     *     A new UserData object derived from the data contained within the
+     *     given Credentials, or null if no such data is present or if the data
+     *     present is invalid.
+     */
+    public UserData fromCredentials(Credentials credentials) {
+
+//        String json;
+//        byte[] correctSignature;
```

Comment:
No commented out code please - if this function is unused, it should be removed.

Reply:
Oh, thanks for pointing that out. It has been removed.